### ISSUE TITLE
Cannot uncheck checkbox field mapped to the entity

### ISSUE THREAD

**Issue by ttibensky** [symfony/symfony#20179]
I'm having issues submitting a form which contains checkbox field. When the mapped field in the entity is true, checkbox is checked and I want to uncheck it and submit the form it won't pass through form validation. I found some issues mentioned this and Im aware of if checkbox is unchecked, its value isn't present in the POST request, so it should be interpreted as null. I have other checkboxes in other forms on my site, but all of the others works. So I was thinking and realized that the form witch doesn't work contains only the checkbox field and nothing else:

``` php
<?php

namespace SomeCompany\SomeBundle\Form\Type\Person\Author;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
use Symfony\Component\Form\Extension\Core\Type\HiddenType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class AlertWhenWrittenType extends AbstractType
{
    /**
     * @param FormBuilderInterface $builder
     * @param array $options
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder->add('checkbox', CheckboxType::class, [
            'required' => false,
        ]);
    }

    /**
     * @param OptionsResolver $resolver
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(['csrf_protection' => false]);
    }
}
```

So I tried to add another field (hidden, not mapped and with empty_data set to true) just to see what happens:

``` php
<?php

namespace SomeCompany\SomeBundle\Form\Type\Person\Author;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
use Symfony\Component\Form\Extension\Core\Type\HiddenType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class AlertWhenWrittenType extends AbstractType
{
    /**
     * @param FormBuilderInterface $builder
     * @param array $options
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder->add('checkbox', CheckboxType::class, [
            'required' => false,
        ]);

        // if there is only a checkbox field in the form,
        // it can't be unchecked, if you uncheck it and submit the form
        // form validation won't pass
        // magically, if I add any another field,
        // you will be able to uncheck the checkbox and validation will pass
        // this is needed until it is fixed in symfony core
        $builder->add('uglyHack', HiddenType::class, [
            'required' => false,
            'mapped' => false,
            'empty_data' => true,
        ]);
    }

    /**
     * @param OptionsResolver $resolver
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(['csrf_protection' => false]);
    }
}
```

also needed to add that field to my twig template

``` twig
{{ form_widget(form.checkbox) }}
{{ form_widget(form.uglyHack) }}
```

and now, miraculously, Im able to uncheck the checkbox and submit the form. Validation now passes and entity field value is now set to false.
I think that its happening because there is only the checkbox field in the form. If checkbox is not checked, its value is not send through POST, but there are no other fields so $_POST is empty
and I think that might cause the problem to appear. Maybe empty $_POST is handled wrong?

I'm using symfony 3.0.9

**Comment by xabbuh** [comment]
The issue is that the Form component is not able to detect that the form was submitted as none of the fields is present (at least when using `handleRequest()`). Though you can workaround that by always calling `submit()` when the request is a POST request.

**Comment by gadelat** [comment]
This reminds me of https://github.com/symfony/symfony/issues/17899

Maybe Symfony forms could set initial state for all the checkboxes (in case of POST) as boolean false and then continue to work with checkboxes internally as booleans, instead of expecting nulls. Both problems would be solved.

**Comment by ttibensky** [comment]
@gadelat but thats not how html checkbox works. I agree with what you said but it would need to become a `BooleanType`. Because `CheckboxType` is working exactly like designed = exactly like it should work.

**Comment by gadelat** [comment]
@ttibensky I'm aware that this mess is caused by how forms in web browsers were designed (wrong IMO), but it needs to be dealt with somehow.

**Comment by backbone87** [comment]
in Contao this issue is solved by including a hidden input in front of the checkbox input with the same name as the checkbox, this way if the checkbox is unchecked the POST data looks like `CHECKBOX_NAME=` and if its checked `CHECKBOX_NAME=&CHECKBOX_NAME=CHECKBOX_VALUE` and because how PHP parses POST data the empty value gets overridden with `CHECKBOX_VALUE` in the latter case.

**Comment by xabbuh** [comment]
see #20210 for a PR

**Comment by ttibensky** [comment]
@xabbuh looks good, this solves the issue perfectly. Feel free to close this ticket after it is merged.

**Comment by HeahDude** [comment]
@ttibensky Could you please confirm that https://github.com/symfony/symfony/pull/17771 is fixing your issue as well? Thanks!

**Comment by ttibensky** [comment]
@HeahDude I'm not using PATCH method, but yeah, it does the thing. For me this ticket is done.

**Comment by HeahDude** [comment]
Thank you @ttibensky for confirming this :)

**Comment by xabbuh** [comment]
I just noticed that I closed here accidentally without noticing that #17771 was never merged. However, we can still keep it closed as we also have #14938 which (basically) covers the same issue.
